Index: json_object_private.h
===================================================================
--- json_object_private.h	(revision 21)
+++ json_object_private.h	(working copy)
@@ -24,7 +24,7 @@
   int _ref_count;
   struct printbuf *_pb;
   union data {
-    boolean c_boolean;
+    json_boolean c_boolean;
     double c_double;
     int c_int;
     struct lh_table *c_object;
Index: json_object.c
===================================================================
--- json_object.c	(revision 21)
+++ json_object.c	(working copy)
@@ -28,7 +28,7 @@
   char* strndup(const char* str, size_t n);
 #endif /* !HAVE_STRNDUP */
 
-#define REFCOUNT_DEBUG 1
+//#define REFCOUNT_DEBUG 0
 
 char *json_number_chars = "0123456789.+-eE";
 char *json_hex_chars = "0123456789abcdef";
@@ -182,7 +182,7 @@
 
 enum json_type json_object_get_type(struct json_object *this)
 {
-  return this->o_type;
+	return (this ? this->o_type : json_type_null);
 }
 
 
@@ -286,7 +286,7 @@
   else return sprintbuf(pb, "false");
 }
 
-struct json_object* json_object_new_boolean(boolean b)
+struct json_object* json_object_new_boolean(json_boolean b)
 {
   struct json_object *this = json_object_new(json_type_boolean);
   if(!this) return NULL;
@@ -295,7 +295,7 @@
   return this;
 }
 
-boolean json_object_get_boolean(struct json_object *this)
+json_boolean json_object_get_boolean(struct json_object *this)
 {
   if(!this) return FALSE;
   switch(this->o_type) {
Index: json_object.h
===================================================================
--- json_object.h	(revision 21)
+++ json_object.h	(working copy)
@@ -15,17 +15,17 @@
 #define JSON_OBJECT_DEF_HASH_ENTIRES 16
 
 #undef FALSE
-#define FALSE ((boolean)0)
+#define FALSE ((json_boolean)0)
 
 #undef TRUE
-#define TRUE ((boolean)1)
+#define TRUE ((json_boolean)1)
 
 extern char *json_number_chars;
 extern char *json_hex_chars;
 
 /* forward structure definitions */
 
-typedef int boolean;
+typedef int json_boolean;
 struct printbuf;
 struct lh_table;
 struct array_list;
@@ -225,7 +225,7 @@
  * @param b a boolean TRUE or FALSE (0 or 1)
  * @returns a json_object of type json_type_boolean
  */
-extern struct json_object* json_object_new_boolean(boolean b);
+extern struct json_object* json_object_new_boolean(json_boolean b);
 
 /** Get the boolean value of a json_object
  *
@@ -236,9 +236,9 @@
  * TRUE will be returned if the object is not NULL.
  *
  * @param obj the json_object instance
- * @returns a boolean
+ * @returns a json_boolean
  */
-extern boolean json_object_get_boolean(struct json_object *obj);
+extern json_boolean json_object_get_boolean(struct json_object *obj);
 
 
 /* int type methods */
Index: debug.h
===================================================================
--- debug.h	(revision 21)
+++ debug.h	(working copy)
@@ -39,4 +39,9 @@
 #define MC_INFO(x, ...) if (0) mc_info(x, ##__VA_ARGS__)
 #endif
 
+/// TODO Check if we can move this somewhere else.
+#ifndef __GNUC__
+#	define __attribute__(x) /* NOTHING My friends */
 #endif
+
+#endif
Index: test1.c
===================================================================
--- test1.c	(revision 21)
+++ test1.c	(working copy)
@@ -11,9 +11,15 @@
   struct json_object *my_string, *my_int, *my_object, *my_array;
   struct json_object *new_obj;
   int i;
-
+  
+  puts("Starting JSON test.");
+  
   MC_SET_DEBUG(1);
 
+  new_obj = json_tokener_parse("null");
+  json_object_get_type(new_obj);
+  json_object_put(new_obj);
+
   my_string = json_object_new_string("\t");
   printf("my_string=%s\n", json_object_get_string(my_string));
   printf("my_string.to_string()=%s\n", json_object_to_json_string(my_string));
@@ -54,9 +60,9 @@
   json_object_object_del(my_object, "baz");
   /*json_object_object_add(my_object, "arr", my_array);*/
   printf("my_object=\n");
-  json_object_object_foreach(my_object, key, val) {
+  /*json_object_object_foreach(my_object, key, val) {
     printf("\t%s: %s\n", key, json_object_to_json_string(val));
-  }
+  }*/
   printf("my_object.to_string()=%s\n", json_object_to_json_string(my_object));
 
   new_obj = json_tokener_parse("\"\003\"");
Index: json-c.vcproj
===================================================================
--- json-c.vcproj	(revision 21)
+++ json-c.vcproj	(working copy)
@@ -1,179 +1,257 @@
-<?xml version="1.0" encoding="Windows-1252"?>
-<VisualStudioProject
-	ProjectType="Visual C++"
-	Version="7.10"
-	Name="json-c"
-	ProjectGUID="{04D8CDBE-FB3E-4362-87E6-07DC3C0083B2}"
-	Keyword="Win32Proj">
-	<Platforms>
-		<Platform
-			Name="Win32"/>
-	</Platforms>
-	<Configurations>
-		<Configuration
-			Name="Debug|Win32"
-			OutputDirectory="Debug"
-			IntermediateDirectory="Debug"
-			ConfigurationType="4"
-			CharacterSet="2">
-			<Tool
-				Name="VCCLCompilerTool"
-				Optimization="0"
-				PreprocessorDefinitions="WIN32;_DEBUG;_LIB"
-				MinimalRebuild="TRUE"
-				BasicRuntimeChecks="3"
-				RuntimeLibrary="5"
-				UsePrecompiledHeader="0"
-				WarningLevel="3"
-				Detect64BitPortabilityProblems="TRUE"
-				DebugInformationFormat="4"/>
-			<Tool
-				Name="VCCustomBuildTool"/>
-			<Tool
-				Name="VCLibrarianTool"
-				OutputFile="$(OutDir)/jsonD.lib"/>
-			<Tool
-				Name="VCMIDLTool"/>
-			<Tool
-				Name="VCPostBuildEventTool"/>
-			<Tool
-				Name="VCPreBuildEventTool"
-				CommandLine="copy config.h.win32 config.h"/>
-			<Tool
-				Name="VCPreLinkEventTool"/>
-			<Tool
-				Name="VCResourceCompilerTool"/>
-			<Tool
-				Name="VCWebServiceProxyGeneratorTool"/>
-			<Tool
-				Name="VCXMLDataGeneratorTool"/>
-			<Tool
-				Name="VCManagedWrapperGeneratorTool"/>
-			<Tool
-				Name="VCAuxiliaryManagedWrapperGeneratorTool"/>
-		</Configuration>
-		<Configuration
-			Name="Release|Win32"
-			OutputDirectory="Release"
-			IntermediateDirectory="Release"
-			ConfigurationType="4"
-			CharacterSet="2">
-			<Tool
-				Name="VCCLCompilerTool"
-				PreprocessorDefinitions="WIN32;NDEBUG;_LIB"
-				RuntimeLibrary="4"
-				UsePrecompiledHeader="0"
-				WarningLevel="3"
-				Detect64BitPortabilityProblems="TRUE"
-				DebugInformationFormat="3"/>
-			<Tool
-				Name="VCCustomBuildTool"/>
-			<Tool
-				Name="VCLibrarianTool"
-				OutputFile="$(OutDir)/json.lib"/>
-			<Tool
-				Name="VCMIDLTool"/>
-			<Tool
-				Name="VCPostBuildEventTool"/>
-			<Tool
-				Name="VCPreBuildEventTool"/>
-			<Tool
-				Name="VCPreLinkEventTool"/>
-			<Tool
-				Name="VCResourceCompilerTool"/>
-			<Tool
-				Name="VCWebServiceProxyGeneratorTool"/>
-			<Tool
-				Name="VCXMLDataGeneratorTool"/>
-			<Tool
-				Name="VCManagedWrapperGeneratorTool"/>
-			<Tool
-				Name="VCAuxiliaryManagedWrapperGeneratorTool"/>
-		</Configuration>
-	</Configurations>
-	<References>
-	</References>
-	<Files>
-		<Filter
-			Name="Source Files"
-			Filter="cpp;c;cxx;def;odl;idl;hpj;bat;asm;asmx"
-			UniqueIdentifier="{4FC737F1-C7A5-4376-A066-2A32D752A2FF}">
-			<File
-				RelativePath=".\arraylist.c">
-			</File>
-			<File
-				RelativePath=".\debug.c">
-			</File>
-			<File
-				RelativePath=".\json_object.c">
-			</File>
-			<File
-				RelativePath=".\json_tokener.c">
-			</File>
-			<File
-				RelativePath=".\json_util.c">
-			</File>
-			<File
-				RelativePath=".\linkhash.c">
-			</File>
-			<File
-				RelativePath=".\printbuf.c">
-			</File>
-		</Filter>
-		<Filter
-			Name="Header Files"
-			Filter="h;hpp;hxx;hm;inl;inc;xsd"
-			UniqueIdentifier="{93995380-89BD-4b04-88EB-625FBE52EBFB}">
-			<File
-				RelativePath=".\arraylist.h">
-			</File>
-			<File
-				RelativePath=".\bits.h">
-			</File>
-			<File
-				RelativePath=".\debug.h">
-			</File>
-			<File
-				RelativePath=".\json_object.h">
-			</File>
-			<File
-				RelativePath=".\json_object_private.h">
-			</File>
-			<File
-				RelativePath=".\json_tokener.h">
-			</File>
-			<File
-				RelativePath=".\json_util.h">
-			</File>
-			<File
-				RelativePath=".\linkhash.h">
-			</File>
-			<File
-				RelativePath=".\printbuf.h">
-			</File>
-		</Filter>
-		<Filter
-			Name="Resource Files"
-			Filter="rc;ico;cur;bmp;dlg;rc2;rct;bin;rgs;gif;jpg;jpeg;jpe;resx"
-			UniqueIdentifier="{67DA6AB6-F800-4c08-8B7A-83BB121AAD01}">
-		</Filter>
-		<Filter
-			Name="Documentation"
-			Filter="">
-			<File
-				RelativePath=".\Doxyfile">
-			</File>
-		</Filter>
-		<File
-			RelativePath=".\config.h.win32">
-		</File>
-		<File
-			RelativePath=".\README-WIN32.html">
-		</File>
-		<File
-			RelativePath=".\README.html">
-		</File>
-	</Files>
-	<Globals>
-	</Globals>
-</VisualStudioProject>
+<?xml version="1.0" encoding="Windows-1252"?>
+<VisualStudioProject
+	ProjectType="Visual C++"
+	Version="8.00"
+	Name="json-c"
+	ProjectGUID="{04D8CDBE-FB3E-4362-87E6-07DC3C0083B2}"
+	RootNamespace="json-c"
+	Keyword="Win32Proj"
+	>
+	<Platforms>
+		<Platform
+			Name="Win32"
+		/>
+	</Platforms>
+	<ToolFiles>
+	</ToolFiles>
+	<Configurations>
+		<Configuration
+			Name="Debug|Win32"
+			OutputDirectory="Debug"
+			IntermediateDirectory="Debug"
+			ConfigurationType="4"
+			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
+			CharacterSet="2"
+			>
+			<Tool
+				Name="VCPreBuildEventTool"
+				CommandLine=""
+			/>
+			<Tool
+				Name="VCCustomBuildTool"
+			/>
+			<Tool
+				Name="VCXMLDataGeneratorTool"
+			/>
+			<Tool
+				Name="VCWebServiceProxyGeneratorTool"
+			/>
+			<Tool
+				Name="VCMIDLTool"
+			/>
+			<Tool
+				Name="VCCLCompilerTool"
+				Optimization="0"
+				PreprocessorDefinitions="WIN32;_DEBUG;_LIB"
+				MinimalRebuild="true"
+				BasicRuntimeChecks="3"
+				RuntimeLibrary="3"
+				UsePrecompiledHeader="0"
+				WarningLevel="3"
+				Detect64BitPortabilityProblems="true"
+				DebugInformationFormat="4"
+			/>
+			<Tool
+				Name="VCManagedResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCPreLinkEventTool"
+			/>
+			<Tool
+				Name="VCLibrarianTool"
+				OutputFile="$(OutDir)/jsonD.lib"
+			/>
+			<Tool
+				Name="VCALinkTool"
+			/>
+			<Tool
+				Name="VCXDCMakeTool"
+			/>
+			<Tool
+				Name="VCBscMakeTool"
+			/>
+			<Tool
+				Name="VCFxCopTool"
+			/>
+			<Tool
+				Name="VCPostBuildEventTool"
+			/>
+		</Configuration>
+		<Configuration
+			Name="Release|Win32"
+			OutputDirectory="Release"
+			IntermediateDirectory="Release"
+			ConfigurationType="4"
+			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
+			CharacterSet="2"
+			>
+			<Tool
+				Name="VCPreBuildEventTool"
+			/>
+			<Tool
+				Name="VCCustomBuildTool"
+			/>
+			<Tool
+				Name="VCXMLDataGeneratorTool"
+			/>
+			<Tool
+				Name="VCWebServiceProxyGeneratorTool"
+			/>
+			<Tool
+				Name="VCMIDLTool"
+			/>
+			<Tool
+				Name="VCCLCompilerTool"
+				PreprocessorDefinitions="WIN32;NDEBUG;_LIB"
+				RuntimeLibrary="2"
+				UsePrecompiledHeader="0"
+				WarningLevel="3"
+				Detect64BitPortabilityProblems="true"
+				DebugInformationFormat="3"
+			/>
+			<Tool
+				Name="VCManagedResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCPreLinkEventTool"
+			/>
+			<Tool
+				Name="VCLibrarianTool"
+				OutputFile="$(OutDir)/json.lib"
+			/>
+			<Tool
+				Name="VCALinkTool"
+			/>
+			<Tool
+				Name="VCXDCMakeTool"
+			/>
+			<Tool
+				Name="VCBscMakeTool"
+			/>
+			<Tool
+				Name="VCFxCopTool"
+			/>
+			<Tool
+				Name="VCPostBuildEventTool"
+			/>
+		</Configuration>
+	</Configurations>
+	<References>
+	</References>
+	<Files>
+		<Filter
+			Name="Source Files"
+			Filter="cpp;c;cxx;def;odl;idl;hpj;bat;asm;asmx"
+			UniqueIdentifier="{4FC737F1-C7A5-4376-A066-2A32D752A2FF}"
+			>
+			<File
+				RelativePath=".\arraylist.c"
+				>
+			</File>
+			<File
+				RelativePath=".\debug.c"
+				>
+			</File>
+			<File
+				RelativePath=".\json_object.c"
+				>
+			</File>
+			<File
+				RelativePath=".\json_tokener.c"
+				>
+			</File>
+			<File
+				RelativePath=".\json_util.c"
+				>
+			</File>
+			<File
+				RelativePath=".\linkhash.c"
+				>
+			</File>
+			<File
+				RelativePath=".\printbuf.c"
+				>
+			</File>
+		</Filter>
+		<Filter
+			Name="Header Files"
+			Filter="h;hpp;hxx;hm;inl;inc;xsd"
+			UniqueIdentifier="{93995380-89BD-4b04-88EB-625FBE52EBFB}"
+			>
+			<File
+				RelativePath=".\arraylist.h"
+				>
+			</File>
+			<File
+				RelativePath=".\bits.h"
+				>
+			</File>
+			<File
+				RelativePath=".\config.h"
+				>
+			</File>
+			<File
+				RelativePath=".\debug.h"
+				>
+			</File>
+			<File
+				RelativePath=".\json_object.h"
+				>
+			</File>
+			<File
+				RelativePath=".\json_object_private.h"
+				>
+			</File>
+			<File
+				RelativePath=".\json_tokener.h"
+				>
+			</File>
+			<File
+				RelativePath=".\json_util.h"
+				>
+			</File>
+			<File
+				RelativePath=".\linkhash.h"
+				>
+			</File>
+			<File
+				RelativePath=".\printbuf.h"
+				>
+			</File>
+		</Filter>
+		<Filter
+			Name="Resource Files"
+			Filter="rc;ico;cur;bmp;dlg;rc2;rct;bin;rgs;gif;jpg;jpeg;jpe;resx"
+			UniqueIdentifier="{67DA6AB6-F800-4c08-8B7A-83BB121AAD01}"
+			>
+		</Filter>
+		<Filter
+			Name="Documentation"
+			>
+			<File
+				RelativePath=".\Doxyfile"
+				>
+			</File>
+		</Filter>
+		<File
+			RelativePath=".\config.h.win32"
+			>
+		</File>
+		<File
+			RelativePath=".\README-WIN32.html"
+			>
+		</File>
+		<File
+			RelativePath=".\README.html"
+			>
+		</File>
+	</Files>
+	<Globals>
+	</Globals>
+</VisualStudioProject>
